File: leetcode/lc799.py
class Solution:
    def champagneTower(self, poured, query_row, query_glass):
        """
        :type poured: int
        :type query_row: int
        :type query_glass: int
        :rtype: float
        """
        memory = {}

        # Return (in-glass, excessive)
        def search(row, glass):
            try:
                return memory[(row, glass)]
            except KeyError:
                pass
            if glass < 0 or glass > row:
                return (0, 0)
            if row == 0 and glass == 0:
                rv = (min(1, poured), max(poured-1, 0))
            else:
                in_ = search(row-1, glass-1)[1] / 2 + search(row-1, glass)[1] / 2
                rv = (min(1, in_), max(in_-1, 0))

            memory[(row, glass)] = rv
            return rv

        return search(query_row, query_glass)[0]

File: leetcode/lc799.cpp
#include <iostream>
#include <cstring>
#include <algorithm>
#include <utility>

using namespace std;

class Solution {
public:
    double champagneTower(int poured_, int query_row, int query_glass) {
        memset(inited, 0, sizeof(inited));
        poured = poured_;

        return search(query_row, query_glass).first;
    }

    std::pair<double, double> search(int row, int glass) {
        if (glass < 0 || glass > row)
        {
            return {0, 0};
        }
        if (inited[row][glass])
        {
            return inglass_and_excessive[row][glass];
        }
        std::pair<double, double>& rv = inglass_and_excessive[row][glass];
        if (row == 0 && glass == 0)
        {
            rv = {std::min(1, poured), std::max(poured-1, 0)};
        }
        else
        {
            double in = (search(row-1, glass-1).second + search(row-1, glass).second)/2;
            rv = {std::min(1., in), std::max(in-1, 0.)};
        }
        //cout << "(" << row << ", " << glass << ")=(" << rv.first << ", " << rv.second << ")" << endl;
        inited[row][glass] = true;
        return rv;
    }

private:
    std::pair<double, double> inglass_and_excessive[100][100];
    bool inited[100][100];
    int poured;
};


int main(void)
{
    Solution().champagneTower(100, 85, 39);

    return 0;
}
